File: app/frontend/src/core/services/__tests__/BaseService.test.ts
import { BaseService } from '../BaseService';
import { BaseEntity } from '../../interfaces/BaseEntity';
import { Repository, RepositoryResponse } from '../../interfaces/Repository';
import { AppError } from '../../utils/errors';
import { ID, PaginationParams, SortParams, QueryFilters } from '../../types/common';

// Mock entity for testing
interface TestEntity extends BaseEntity {
  name: string;
  value: number;
}

// Mock repository implementation
class MockRepository implements Repository<TestEntity> {
  private entities: TestEntity[] = [];

  async findById(id: ID): Promise<TestEntity | null> {
    return this.entities.find(e => e.id === id) || null;
  }

  async findAll(
    pagination?: PaginationParams,
    sort?: SortParams,
    filters?: QueryFilters
  ): Promise<RepositoryResponse<TestEntity>> {
    let result = [...this.entities];

    // Apply filters if provided
    if (filters) {
      result = result.filter(entity => {
        return Object.entries(filters).every(([key, value]) => {
          return entity[key as keyof TestEntity] === value;
        });
      });
    }

    // Apply sorting if provided
    if (sort) {
      result.sort((a, b) => {
        const aValue = a[sort.field as keyof TestEntity];
        const bValue = b[sort.field as keyof TestEntity];
        if (aValue === undefined || bValue === undefined) return 0;
        return sort.order === 'asc' ? 
          (aValue < bValue ? -1 : 1) : 
          (aValue > bValue ? -1 : 1);
      });
    }

    // Apply pagination
    const { page = 1, limit = 10 } = pagination || {};
    const start = (page - 1) * limit;
    const paginatedData = result.slice(start, start + limit);

    return {
      data: paginatedData,
      total: result.length
    };
  }

  async create(entity: TestEntity): Promise<TestEntity> {
    this.entities.push(entity);
    return entity;
  }

  async update(id: ID, entity: TestEntity): Promise<TestEntity | null> {
    const index = this.entities.findIndex(e => e.id === id);
    if (index === -1) return null;
    this.entities[index] = entity;
    return entity;
  }

  async delete(id: ID): Promise<boolean> {
    const index = this.entities.findIndex(e => e.id === id);
    if (index === -1) return false;
    this.entities.splice(index, 1);
    return true;
  }
}

// Concrete implementation of BaseService for testing
class TestService extends BaseService<TestEntity> {
  constructor(repository: Repository<TestEntity>) {
    super(repository);
  }
}

describe('BaseService', () => {
  let service: TestService;
  let repository: MockRepository;

  beforeEach(() => {
    repository = new MockRepository();
    service = new TestService(repository);
  });

  describe('findById', () => {
    it('should find entity by ID', async () => {
      const entity: TestEntity = {
        id: 1,
        name: 'Test',
        value: 42,
        createdAt: new Date(),
        updatedAt: new Date()
      };
      await repository.create(entity);

      const result = await service.findById(1);
      expect(result).toEqual(entity);
    });

    it('should return null for non-existent ID', async () => {
      const result = await service.findById(999);
      expect(result).toBeNull();
    });

    it('should handle errors', async () => {
      jest.spyOn(repository, 'findById').mockRejectedValue(new Error('DB Error'));
      await expect(service.findById(1)).rejects.toThrow(AppError);
    });
  });

  describe('findAll', () => {
    beforeEach(async () => {
      await repository.create({
        id: 1,
        name: 'First',
        value: 10,
        createdAt: new Date(),
        updatedAt: new Date()
      });
      await repository.create({
        id: 2,
        name: 'Second',
        value: 20,
        createdAt: new Date(),
        updatedAt: new Date()
      });
    });

    it('should return paginated results', async () => {
      const result = await service.findAll({
        pagination: { page: 1, limit: 1 }
      });

      expect(result.data.length).toBe(1);
      expect(result.meta?.total).toBe(2);
      expect(result.meta?.page).toBe(1);
      expect(result.meta?.limit).toBe(1);
    });

    it('should apply sorting', async () => {
      const result = await service.findAll({
        sort: { field: 'value', order: 'desc' }
      });

      expect(result.data[0].value).toBe(20);
    });

    it('should apply filters', async () => {
      const result = await service.findAll({
        filters: { value: 10 }
      });

      expect(result.data.length).toBe(1);
      expect(result.data[0].name).toBe('First');
    });

    it('should handle errors', async () => {
      jest.spyOn(repository, 'findAll').mockRejectedValue(new Error('DB Error'));
      await expect(service.findAll({})).rejects.toThrow(AppError);
    });
  });

  describe('create', () => {
    it('should create new entity', async () => {
      const newEntity = {
        name: 'New',
        value: 30
      };

      const result = await service.create(newEntity);
      expect(result.name).toBe('New');
      expect(result.createdAt).toBeDefined();
      expect(result.updatedAt).toBeDefined();
    });

    it('should handle errors', async () => {
      jest.spyOn(repository, 'create').mockRejectedValue(new Error('DB Error'));
      await expect(service.create({ name: 'Test', value: 1 })).rejects.toThrow(AppError);
    });
  });

  describe('update', () => {
    it('should update existing entity', async () => {
      const entity = await repository.create({
        id: 1,
        name: 'Original',
        value: 10,
        createdAt: new Date(),
        updatedAt: new Date()
      });

      const result = await service.update(1, { ...entity, name: 'Updated' });
      expect(result?.name).toBe('Updated');
      expect(result?.updatedAt).toBeDefined();
      const resultTime = result?.updatedAt?.getTime() ?? 0;
      const entityTime = entity.updatedAt?.getTime() ?? 0;
      expect(resultTime).toBeGreaterThan(entityTime);
    });

    it('should return null for non-existent entity', async () => {
      const result = await service.update(999, { name: 'Updated', value: 1 });
      expect(result).toBeNull();
    });

    it('should handle errors', async () => {
      jest.spyOn(repository, 'update').mockRejectedValue(new Error('DB Error'));
      await expect(service.update(1, { name: 'Test', value: 1 })).rejects.toThrow(AppError);
    });
  });

  describe('delete', () => {
    it('should delete existing entity', async () => {
      await repository.create({
        id: 1,
        name: 'ToDelete',
        value: 10,
        createdAt: new Date(),
        updatedAt: new Date()
      });

      const result = await service.delete(1);
      expect(result).toBe(true);
      expect(await service.findById(1)).toBeNull();
    });

    it('should return false for non-existent entity', async () => {
      const result = await service.delete(999);
      expect(result).toBe(false);
    });

    it('should handle errors', async () => {
      jest.spyOn(repository, 'delete').mockRejectedValue(new Error('DB Error'));
      await expect(service.delete(1)).rejects.toThrow(AppError);
    });
  });

  describe('exists', () => {
    it('should return true for existing entity', async () => {
      await repository.create({
        id: 1,
        name: 'Test',
        value: 10,
        createdAt: new Date(),
        updatedAt: new Date()
      });

      const result = await service.exists(1);
      expect(result).toBe(true);
    });

    it('should return false for non-existent entity', async () => {
      const result = await service.exists(999);
      expect(result).toBe(false);
    });

    it('should handle errors', async () => {
      jest.spyOn(repository, 'findById').mockRejectedValue(new Error('DB Error'));
      await expect(service.exists(1)).rejects.toThrow(AppError);
    });
  });
}); 

File: backend/app/src/core/base/__tests__/base.service.test.py
from typing import Any, Dict, List


  ServiceResponse,
  ServiceConfig,
  PaginatedResponse,
  ValidationResult,
  QueryParams,
  PaginationParams,
  ID
} from '../types/common'
class TestEntity:
    name: str
    value: float
const testValidationSchema = z.object({
  name: z.string().min(1, 'Name is required and must be a string'),
  value: z.number().min(0, 'Value must be a non-negative number')
})
TestValidationSchema = z.infer<typeof testValidationSchema>
class TestService extends ValidatableService<TestEntity> {
  private entities: Map<ID, TestEntity> = new Map()
  private nextId = 1
  constructor() {
    super()
    this.initializeValidationRules()
  }
  private initializeValidationRules() {
    this.addValidationRule('name', {
      required: true,
      minLength: 1
    })
    this.addValidationRule('value', {
      required: true,
      min: 0
    })
  }
  protected async applyValidationRule(
    field: keyof TestEntity,
    value: Any,
    ruleName: str,
    ruleConfig: Any
  ): Promise<boolean> {
    switch (ruleName) {
      case 'required':
        return value !== undefined && value !== null && value !== ''
      case 'minLength':
        return typeof value === 'string' && value.length >= ruleConfig
      case 'min':
        return typeof value === 'number' && value >= ruleConfig
      default:
        return true
    }
  }
  protected getValidationMessage(
    field: keyof TestEntity,
    ruleName: str,
    ruleConfig: Any
  ): str {
    switch (ruleName) {
      case 'required':
        return `${String(field)} is required`
      case 'minLength':
        return `${String(field)} must be at least ${ruleConfig} characters long`
      case 'min':
        return `${String(field)} must be at least ${ruleConfig}`
      default:
        return `Invalid ${String(field)}`
    }
  }
  protected async executeCreate(data: Omit<TestEntity, keyof BaseEntity>): Promise<TestEntity> {
    const id = this.nextId++
    const entity: \'TestEntity\' = {
      id,
      createdAt: new Date(),
      updatedAt: new Date(),
      ...data
    }
    this.entities.set(id, entity)
    return entity
  }
  protected async executeUpdate(id: ID, data: Partial<TestEntity>): Promise<TestEntity> {
    const entity = this.entities.get(id)
    if (!entity) {
      throw new AppError('Entity not found', 404)
    }
    const updatedEntity: \'TestEntity\' = {
      ...entity,
      ...data,
      updatedAt: new Date()
    }
    this.entities.set(id, updatedEntity)
    return updatedEntity
  }
  protected async executeDelete(id: ID): Promise<void> {
    if (!this.entities.has(id)) {
      throw new AppError('Entity not found', 404)
    }
    this.entities.delete(id)
  }
  protected async executeFindById(id: ID): Promise<TestEntity> {
    const entity = this.entities.get(id)
    if (!entity) {
      throw new AppError('Entity not found', 404)
    }
    return entity
  }
  /**
   * Execute find one operation
   */
  protected async executeFindOne(params: QueryParams<TestEntity>): Promise<TestEntity | null> {
    const entities = Array.from(this.entities.values())
    if (params.filter) {
      const filtered = entities.filter(entity => {
        return Object.entries(params.filter as Partial<TestEntity>).every(([key, value]) => {
          const entityValue = entity[key as keyof TestEntity]
          return entityValue === value
        })
      })
      return filtered[0] || null
    }
    return entities[0] || null
  }
  /**
   * Execute find all operation
   */
  protected async executeFindAll(params?: QueryParams<TestEntity> & PaginationParams): Promise<PaginatedResponse<TestEntity>> {
    let entities = Array.from(this.entities.values())
    if (params?.filter) {
      entities = entities.filter(entity => {
        return Object.entries(params.filter as Partial<TestEntity>).every(([key, value]) => {
          const entityValue = entity[key as keyof TestEntity]
          return entityValue === value
        })
      })
    }
    if (params?.search) {
      const searchStr = params.search.toLowerCase()
      entities = entities.filter(entity =>
        entity.name.toLowerCase().includes(searchStr)
      )
    }
    if (params?.sort) {
      const sortEntries = Object.entries(params.sort)
      if (sortEntries.length > 0) {
        const [field, order] = sortEntries[0] as [keyof TestEntity, 'asc' | 'desc']
        entities.sort((a, b) => {
          const aValue = a[field]
          const bValue = b[field]
          if (typeof aValue === 'number' && typeof bValue === 'number') {
            return order === 'desc' ? bValue - aValue : aValue - bValue
          }
          if (typeof aValue === 'string' && typeof bValue === 'string') {
            return order === 'desc' ? 
              bValue.localeCompare(aValue) :
              aValue.localeCompare(bValue)
          }
          const aStr = String(aValue)
          const bStr = String(bValue)
          return order === 'desc' ? 
            bStr.localeCompare(aStr) :
            aStr.localeCompare(bStr)
        })
      }
    }
    const page = params?.page || 1
    const limit = params?.limit || 10
    const start = (page - 1) * limit
    const end = start + limit
    const items = entities.slice(start, end)
    return {
      items,
      total: entities.length,
      page,
      limit,
      hasMore: end < entities.length
    }
  }
  /**
   * Execute validate operation
   */
  protected async executeValidate(data: Partial<TestEntity>): Promise<ValidationResult> {
    try {
      await testValidationSchema.parseAsync(data)
      return { isValid: true, errors: [] }
    } catch (error) {
      if (error instanceof z.ZodError) {
        return {
          isValid: false,
          errors: error.errors.map(err => ({
            field: err.path.join('.'),
            message: err.message,
            code: 'VALIDATION_ERROR'
          }))
        }
      }
      return {
        isValid: false,
        errors: [{
          field: '*',
          message: 'Validation failed unexpectedly',
          code: 'VALIDATION_ERROR'
        }]
      }
    }
  }
  /**
   * Execute validate field operation
   */
  protected async executeValidateField(field: keyof TestEntity, value: Any): Promise<ValidationResult> {
    try {
      const fieldSchema = testValidationSchema.shape[field as keyof TestValidationSchema]
      if (!fieldSchema) {
        return {
          isValid: false,
          errors: [{
            field: String(field),
            message: 'Field does not exist in validation schema',
            code: 'VALIDATION_ERROR'
          }]
        }
      }
      await fieldSchema.parseAsync(value)
      return { isValid: true, errors: [] }
    } catch (error) {
      if (error instanceof z.ZodError) {
        return {
          isValid: false,
          errors: error.errors.map(err => ({
            field: String(field),
            message: err.message,
            code: 'VALIDATION_ERROR'
          }))
        }
      }
      return {
        isValid: false,
        errors: [{
          field: String(field),
          message: 'Field validation failed unexpectedly',
          code: 'VALIDATION_ERROR'
        }]
      }
    }
  }
  /**
   * Get default filterable fields
   */
  protected getDefaultFilterableFields(): Array<keyof TestEntity> {
    return ['name', 'value', 'createdAt', 'updatedAt']
  }
  /**
   * Get default sortable fields
   */
  protected getDefaultSortableFields(): Array<keyof TestEntity> {
    return ['name', 'value', 'createdAt', 'updatedAt']
  }
  /**
   * Check if entity matches query parameters
   */
  private matchesParams(entity: \'TestEntity\', params: QueryParams<TestEntity>): bool {
    if (!params.filter) {
      return true
    }
    return Object.entries(params.filter).every(([key, value]) => {
      const entityValue = entity[key as keyof TestEntity]
      return entityValue === value
    })
  }
  async createMany(data: Array<Omit<TestEntity, keyof BaseEntity>>): Promise<ServiceResponse<TestEntity[]>> {
    try {
      const validationResults = await Promise.all(
        data.map(item => this.validateCreate(item))
      )
      const validationErrors = validationResults.flatMap((result, index) => 
        result.isValid ? [] : result.errors.map(error => ({
          ...error,
          index,
          data: data[index]
        }))
      )
      if (validationErrors.length > 0) {
        throw new AppError('Validation failed', 400)
      }
      const entities = await Promise.all(
        data.map(item => this.executeCreate(item))
      )
      return {
        success: true,
        data: entities
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error : new Error('Unknown error')
      }
    }
  }
  async updateMany(updates: Array<{ id: ID; data: Partial<TestEntity> }>): Promise<ServiceResponse<TestEntity[]>> {
    try {
      const entities = await Promise.all(
        updates.map(async ({ id }) => {
          const entity = await this.executeFindById(id).catch(() => null)
          if (!entity) {
            throw new AppError(`Entity with id ${id} not found`, 404)
          }
          return entity
        })
      )
      const validationResults = await Promise.all(
        updates.map(({ id, data }) => this.validateUpdate(id, data))
      )
      const validationErrors = validationResults.flatMap((result, index) => 
        result.isValid ? [] : result.errors.map(error => ({
          ...error,
          index,
          data: updates[index].data
        }))
      )
      if (validationErrors.length > 0) {
        throw new AppError('Validation failed', 400)
      }
      const updatedEntities = await Promise.all(
        updates.map(({ id, data }) => this.executeUpdate(id, data))
      )
      return {
        success: true,
        data: updatedEntities
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error : new Error('Unknown error')
      }
    }
  }
  async deleteMany(ids: List[ID]): Promise<ServiceResponse<void>> {
    try {
      await Promise.all(
        ids.map(async id => {
          const entity = await this.executeFindById(id).catch(() => null)
          if (!entity) {
            throw new AppError(`Entity with id ${id} not found`, 404)
          }
          return entity
        })
      )
      await Promise.all(
        ids.map(id => this.executeDelete(id))
      )
      return {
        success: true
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error : new Error('Unknown error')
      }
    }
  }
}
describe('TestService', () => {
  let service: \'TestService\'
  beforeEach(() => {
    service = new TestService()
  })
  describe('validation', () => {
    it('should validate required fields', async () => {
      const result = await service.validateCreate({
        name: '',
        value: 100
      })
      expect(result.isValid).toBe(false)
      expect(result.errors).toHaveLength(2)
      expect(result.errors.map(e => e.field)).toContain('name')
      expect(result.errors.map(e => e.message)).toEqual(
        expect.arrayContaining([
          expect.stringMatching(/required/i),
          expect.stringMatching(/at least 1/i)
        ])
      )
    })
    it('should validate minimum value', async () => {
      const result = await service.validateCreate({
        name: 'Test',
        value: -1
      })
      expect(result.isValid).toBe(false)
      expect(result.errors).toHaveLength(1)
      expect(result.errors[0].field).toBe('value')
      expect(result.errors[0].message).toBe('value must be at least 0')
    })
    it('should pass validation for valid data', async () => {
      const result = await service.validateCreate({
        name: 'Test',
        value: 100
      })
      expect(result.isValid).toBe(true)
      expect(result.errors).toHaveLength(0)
    })
    it('should validate partial updates', async () => {
      const result = await service.validateUpdate(1, {
        name: ''
      })
      expect(result.isValid).toBe(false)
      expect(result.errors).toHaveLength(2)
      expect(result.errors.map(e => e.field)).toContain('name')
      expect(result.errors.map(e => e.message)).toEqual(
        expect.arrayContaining([
          expect.stringMatching(/required/i),
          expect.stringMatching(/at least 1/i)
        ])
      )
    })
  })
  describe('create', () => {
    it('should create a new entity', async () => {
      const data = { name: 'Test', value: 100 }
      const result = await service.create(data)
      expect(result.success).toBe(true)
      expect(result.data).toMatchObject({
        ...data,
        id: expect.any(Number),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date)
      })
    })
    it('should validate data before creation', async () => {
      const data = { name: '', value: -1 }
      const result = await service.create(data)
      expect(result.success).toBe(false)
      expect(result.error).toBeDefined()
      expect(result.error?.message).toContain('Validation failed')
    })
  })
  describe('update', () => {
    let entityId: ID
    beforeEach(async () => {
      const result = await service.create({ name: 'Test', value: 100 })
      entityId = result.data!.id
    })
    it('should update an existing entity', async () => {
      const data = { name: 'Updated', value: 200 }
      const result = await service.update(entityId, data)
      expect(result.success).toBe(true)
      expect(result.data).toMatchObject({
        ...data,
        id: entityId
      })
    })
    it('should validate data before update', async () => {
      const data = { name: '', value: -1 }
      const result = await service.update(entityId, data)
      expect(result.success).toBe(false)
      expect(result.error).toBeDefined()
      expect(result.error?.message).toContain('Validation failed')
    })
    it('should handle non-existent entity', async () => {
      const result = await service.update(999, { name: 'Updated' })
      expect(result.success).toBe(false)
      expect(result.error).toBeDefined()
      expect(result.error?.message).toContain('Validation failed')
    })
  })
  describe('delete', () => {
    let entityId: ID
    beforeEach(async () => {
      const result = await service.create({ name: 'Test', value: 100 })
      entityId = result.data!.id
    })
    it('should delete an existing entity', async () => {
      const result = await service.delete(entityId)
      expect(result.success).toBe(true)
      const findResult = await service.findById(entityId)
      expect(findResult.success).toBe(false)
    })
  })
  describe('findById', () => {
    let entityId: ID
    beforeEach(async () => {
      const result = await service.create({ name: 'Test', value: 100 })
      entityId = result.data!.id
    })
    it('should find an entity by id', async () => {
      const result = await service.findById(entityId)
      expect(result.success).toBe(true)
      expect(result.data).toMatchObject({
        id: entityId,
        name: 'Test',
        value: 100
      })
    })
  })
  describe('findAll', () => {
    beforeEach(async () => {
      await service.create({ name: 'Test 1', value: 100 })
      await service.create({ name: 'Test 2', value: 200 })
      await service.create({ name: 'Test 3', value: 300 })
    })
    it('should return all entities with pagination', async () => {
      const result = await service.findAll({
        page: 1,
        limit: 2
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items).toHaveLength(2)
      expect(result.data?.total).toBe(3)
      expect(result.data?.page).toBe(1)
      expect(result.data?.limit).toBe(2)
      expect(result.data?.hasMore).toBe(true)
    })
    it('should filter entities', async () => {
      const result = await service.findAll({
        filter: Dict[str, Any]
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items).toHaveLength(1)
      expect(result.data?.items[0].value).toBe(200)
    })
    it('should sort entities', async () => {
      const result = await service.findAll({
        sort: Dict[str, Any]
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items[0].value).toBe(300)
      expect(result.data?.items[2].value).toBe(100)
    })
    it('should handle empty filter', async () => {
      const result = await service.findAll({
        filter: {}
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items).toHaveLength(3)
    })
    it('should handle non-existent filter field', async () => {
      const result = await service.findAll({
        filter: Dict[str, Any]
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items).toHaveLength(0)
    })
    it('should handle invalid sort field', async () => {
      const result = await service.findAll({
        sort: Dict[str, Any]
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items).toHaveLength(3)
    })
    it('should handle search parameter', async () => {
      const result = await service.findAll({
        search: 'Test 2'
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items).toHaveLength(1)
      expect(result.data?.items[0].name).toBe('Test 2')
    })
    it('should handle combined filter, sort, and search', async () => {
      const result = await service.findAll({
        filter: Dict[str, Any],
        sort: Dict[str, Any],
        search: 'Test'
      } as QueryParams<TestEntity> & PaginationParams)
      expect(result.success).toBe(true)
      expect(result.data?.items).toHaveLength(1)
      expect(result.data?.items[0].value).toBe(200)
    })
  })
  describe('bulk operations', () => {
    describe('createMany', () => {
      it('should create multiple entities', async () => {
        const entities = [
          { name: 'Bulk 1', value: 100 },
          { name: 'Bulk 2', value: 200 },
          { name: 'Bulk 3', value: 300 }
        ]
        const result = await service.createMany(entities)
        expect(result.success).toBe(true)
        expect(result.data).toHaveLength(3)
        expect(result.data?.map(e => e.name)).toEqual(['Bulk 1', 'Bulk 2', 'Bulk 3'])
      })
      it('should validate all entities before creation', async () => {
        const entities = [
          { name: 'Valid', value: 100 },
          { name: '', value: -1 },
          { name: 'Also Invalid', value: -2 }
        ]
        const result = await service.createMany(entities)
        expect(result.success).toBe(false)
        expect(result.error).toBeDefined()
        expect(result.error?.message).toContain('Validation failed')
      })
    })
    describe('updateMany', () => {
      let entityIds: List[ID]
      beforeEach(async () => {
        const results = await Promise.all([
          service.create({ name: 'Test 1', value: 100 }),
          service.create({ name: 'Test 2', value: 200 }),
          service.create({ name: 'Test 3', value: 300 })
        ])
        entityIds = results.map(r => r.data!.id)
      })
      it('should update multiple entities', async () => {
        const updates = entityIds.map(id => ({
          id,
          data: Dict[str, Any]
        }))
        const result = await service.updateMany(updates)
        expect(result.success).toBe(true)
        expect(result.data).toHaveLength(3)
        expect(result.data?.every(e => e.value === 1000)).toBe(true)
      })
      it('should validate all updates before applying', async () => {
        const updates = entityIds.map(id => ({
          id,
          data: Dict[str, Any]
        }))
        const result = await service.updateMany(updates)
        expect(result.success).toBe(false)
        expect(result.error).toBeDefined()
        expect(result.error?.message).toContain('Validation failed')
      })
      it('should handle non-existent entities', async () => {
        const updates = [
          { id: 999, data: Dict[str, Any] },
          { id: entityIds[0], data: Dict[str, Any] }
        ]
        const result = await service.updateMany(updates)
        expect(result.success).toBe(false)
        expect(result.error).toBeDefined()
        expect(result.error?.message).toContain('not found')
      })
    })
    describe('deleteMany', () => {
      let entityIds: List[ID]
      beforeEach(async () => {
        const results = await Promise.all([
          service.create({ name: 'Test 1', value: 100 }),
          service.create({ name: 'Test 2', value: 200 }),
          service.create({ name: 'Test 3', value: 300 })
        ])
        entityIds = results.map(r => r.data!.id)
      })
      it('should delete multiple entities', async () => {
        const result = await service.deleteMany(entityIds)
        expect(result.success).toBe(true)
        const findResult = await service.findAll()
        expect(findResult.data?.items).toHaveLength(0)
      })
      it('should handle non-existent entities', async () => {
        const result = await service.deleteMany([...entityIds, 999])
        expect(result.success).toBe(false)
        expect(result.error).toBeDefined()
        expect(result.error?.message).toContain('not found')
      })
    })
  })
  describe('findOne', () => {
    beforeEach(async () => {
      await service.create({ name: 'Test 1', value: 100 })
      await service.create({ name: 'Test 2', value: 200 })
    })
    it('should find one entity matching the filter', async () => {
      const result = await service.findOne({
        filter: Dict[str, Any]
      })
      expect(result.success).toBe(true)
      expect(result.data?.value).toBe(200)
    })
  })
})
describe('Boundary and Large Dataset Tests', () => {
  let service: \'TestService\'
  beforeEach(() => {
    service = new TestService()
  })
  it('should allow maximum allowed value for "value"', async () => {
    const max = Number.MAX_SAFE_INTEGER
    const result = await service.create({ name: 'Max', value: max })
    expect(result.success).toBe(true)
    expect(result.data?.value).toBe(max)
  })
  it('should handle pagination at the last page', async () => {
    for (let i = 0; i < 15; i++) {
      await service.create({ name: `Entity ${i + 1}`, value: i })
    }
    const result = await service.findAll({ page: 2, limit: 10 } as any)
    expect(result.success).toBe(true)
    expect(result.data?.items.length).toBe(5)
    expect(result.data?.page).toBe(2)
    expect(result.data?.hasMore).toBe(false)
  })
  it('should handle large datasets (100+ entities) with pagination', async () => {
    for (let i = 0; i < 120; i++) {
      await service.create({ name: `Entity ${i + 1}`, value: i })
    }
    const result = await service.findAll({ page: 5, limit: 25 } as any)
    expect(result.success).toBe(true)
    expect(result.data?.items.length).toBe(20)
    expect(result.data?.total).toBe(120)
    expect(result.data?.page).toBe(5)
    expect(result.data?.hasMore).toBe(false)
  })
  it.skip('should emit events on create/update/delete', () => {
  })
  it.skip('should hit/miss cache appropriately', () => {
  })
}) 